XenModel/Actions/VM/VMMoveAction.cs

```diff
@@ -95,68 +96,77 @@ private void PopulateApiMethodsToRoleCheck()
+                };
+                newVBD.SetIsOwner(oldVBD.GetIsOwner());
+
+                try
                 {
-                    if (!oldVBD.GetIsOwner())
-                        continue;
-
-                    var curVdi = Connection.Resolve(oldVBD.VDI);
-                    if (curVdi == null)
-                        continue;
-
-                    if (StorageMapping == null || !StorageMapping.ContainsKey(oldVBD.VDI.opaque_ref))
-                        continue;
-
-                    SR sr = StorageMapping[oldVBD.VDI.opaque_ref];
-                    if (sr == null || curVdi.SR.opaque_ref == sr.opaque_ref)
-                        continue;
-
-                    RelatedTask = XenAPI.VDI.async_copy(Session, oldVBD.VDI.opaque_ref, sr.opaque_ref);
-                    PollToCompletion(PercentComplete, PercentComplete + halfstep);
-                    var newVDI = Connection.WaitForCache(new XenRef<VDI>(Result));
- 
-                    var newVBD = new VBD
-                                     {
-                                         userdevice = oldVBD.userdevice,
-                                         bootable = oldVBD.bootable,
-                                         mode = oldVBD.mode,
-                                         type = oldVBD.type,
-                                         unpluggable = oldVBD.unpluggable,
-                                         other_config = oldVBD.other_config,
-                                         VDI = new XenRef<VDI>(newVDI.opaque_ref),
-                                         VM = new XenRef<VM>(VM.opaque_ref)
-                                     };
-                    newVBD.SetIsOwner(oldVBD.GetIsOwner());
-
-                    VBD vbd = oldVBD;
-                    BestEffort(ref exn, () => VDI.destroy(Session, vbd.VDI.opaque_ref));
-                    Connection.WaitForCache<VBD>(VBD.create(Session, newVBD));
-
-                    PercentComplete += halfstep;
+                    VDI.destroy(Session, oldVBD.VDI.opaque_ref);
+                }
+                catch (Exception e)
+                {
+                    log.ErrorFormat("Failed to destroy old VDI {0}", oldVBD.VDI.opaque_ref);
+                    exceptions.Add(e);
                 }
 
-                if (SR != null)
-                    VM.set_suspend_SR(Session, VM.opaque_ref, SR.opaque_ref);
-
-                if (exn != null)
-                    throw exn;
+                Connection.WaitForCache(VBD.create(Session, newVBD));
 
+                PercentComplete += halfstep;
             }
-            catch (CancelledException)
-            {
-                this.Description = string.Format(Messages.MOVE_CANCELLED, VM.Name());
-                throw;
-            }
-            this.Description = Messages.MOVED;
+
+            if (SR != null)
+                VM.set_suspend_SR(Session, VM.opaque_ref, SR.opaque_ref);
+
+            if (exceptions.Count > 0)

```

Comment:
Purely a style thing but `if (exceptions.Any())` I think looks a bit clearer.

Reply:
Well, it's not just a style thing. If the sizes are not too huge there is no practical difference, but theoretically speaking, `Count > 0` performs better than `Any()` because `exceptions` is a `List<T>`. If it were an `IEnumerable<T>`, `Any()` would be a better choice over `Count() > 0`. Of course, I'm not going to pretend that the usage of the one vs the other has been taken carefully into account throughout XenCenter; we should probably perform an audit at some point because linq is great for conciseness, but has caveats...